### PR TITLE
api: record merging addition

### LINKED ISSUE #61: api: record merging
Add support in deposit record API for merging changes into record. The merging 
deals with the following situation:

1. Deposit (record v1) fetches record (record v1)
2. Deposit is updated (deposit v2 )
3. Record is updated (record v2)
4. Deposit is published (conflict!)

Since the record was updated after the it was loaded into the deposit, the 
deposit will be missing the updates from record v2.

Basic algorithm: 
- Patch from diff of deposit v2 against record v1 (i.e. which changes did the user make)
- Try to apply patch on record v2. If unsuccesful,  bail out (i.e. let user resovle conflicts).

Implementation:

- Merging should be plugged [here](https://github.com/inveniosoftware/invenio-deposit/blob/master/invenio_deposit/api.py#L214-L222)
- Implement algorithm in [``merge_with_published()``](https://github.com/inveniosoftware/invenio-deposit/blob/master/invenio_deposit/api.py#L133)
- [Invenio 2.0 version](https://github.com/inveniosoftware/invenio/blob/maint-2.0/invenio/modules/deposit/tasks.py#L258-L342) of same issue.

### PR DESCRIPTION
**Pull request by hachreak**
* Adds record merging. (closes #61)

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio_deposit/api.py [hunk 1]
```diff
@@ -130,13 +131,21 @@ def fetch_published(self):
         )
         return resolver.resolve(pid_value)
 
-    def merge_with_published(self):
+    def merge_with_published(self, rev_latest):
         """."""
-        # TODO if revisions are the same then we can apply directly
-        # latest = self.fetch_published()
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: Isn't `rev_latest` same as `latest = self.fetch_published()`?

File: invenio_deposit/api.py [hunk 2]
```diff
@@ -130,13 +131,21 @@ def fetch_published(self):
...
-        # initial = A.revisions[self['_deposit']['pid']['revision']]
-        # patch = diff(initial, self)
-        # latest.apply(patch)
+        # the deposit itself
+        rev_now = dict(self.dumps())
+        # check if you need to apply a patch
+        if rev_latest.revision_id == self['_deposit']['pid']['revision_id']:
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: IMHO we don't need to do this check here if we want to force the merge process.

File: invenio_deposit/api.py [hunk 3]
```diff
@@ -211,18 +220,17 @@ def process_files(data):
 
             with process_files(data) as data:
                 record = Record.create(data, id_=id_)
+            self.commit()
+            return self
         else:  # Update after edit
             record_pid, record = self.fetch_published()
-            # TODO add support for patching
-            assert record.revision_id == self['_deposit']['pid']['revision_id']
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```python
if not record.revision_id == self['_deposit']['pid']['revision_id']:
     data = self.merge_with_published(record)
else:
    data = dict(self.dumps())
```

File: invenio_deposit/api.py [hunk 4]
```diff
@@ -211,18 +220,17 @@ def process_files(data):
...
 
-            data = dict(self.dumps())
+            data = dict(new_record.dumps())
             data['$schema'] = self.record_schema
             record = record.__class__(data, model=record.model)
             record.commit()
-
-        self.commit()
-        return self
+            return record
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: please revert this change. You must return deposit (`self`) not record.

File: invenio_deposit/api.py [hunk 5]
```diff
@@ -211,18 +220,17 @@ def process_files(data):
 
             with process_files(data) as data:
                 record = Record.create(data, id_=id_)
+            self.commit()
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: revert this change https://github.com/inveniosoftware/invenio-deposit/pull/63/files#r64379353

File: tests/test_api.py [hunk 6]
```diff
@@ -184,3 +184,45 @@ def test_files_property(app, db, fake_schemas, location):
...
+    #  from sqlalchemy.orm.attributes import flag_modified
+    #  s2 = Session(bind=db.session.connection())
+    #  record = s2.query(RecordMetadata).get(deposit.id)
+    #  record.json['metadata']['poster'] = 'myposter'
+    #  flag_modified(record, 'json')
+    #  import ipdb; ipdb.set_trace()  # XXX BREAKPOINT
+    #  db.session.merge(record)
+    #  import ipdb; ipdb.set_trace()  # XXX BREAKPOINT
+
+    record = Record.get_record(deposit.id)
```
Comment (lnielsen) [inline tests/test_api.py]: ``record = deposit.fetch_published()``

File: invenio_deposit/api.py [hunk 7]
```diff
@@ -131,12 +133,15 @@ def fetch_published(self):
...
-        """."""
-        # TODO if revisions are the same then we can apply directly
-        # latest = self.fetch_published()
-        # initial = A.revisions[self['_deposit']['pid']['revision']]
-        # patch = diff(initial, self)
-        # latest.apply(patch)
+        """Merge changes with latest published version."""
+        latest_id, latest_published = self.fetch_published()
+        current = dict(self.dumps())
+        before = dict(self.revisions[-2].dumps())
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```
pid, published = self.fetch_published()
initial = published.revisions[self['_deposit']['pid']['revision']]
```

Comment (jirikuncar) [inline invenio_deposit/api.py]:
```python

        pid, first = self.fetch_published()
        lca = published.revisions[self['_deposit']['pid']['revision']]
        second = self

        m = Merger(lca.dumps(), first.dumps(), second.dumps(), {})

        try:
            m.run()
        except UnresolvedConflictsException:
            # see https://github.com/inveniosoftware/dictdiffer/blob/master/tests/test_merge.py#L47-L65
```

Comment (jirikuncar) [inline invenio_deposit/api.py]: Please note that we need to take care about `$schema` and `_deposit` fields.

Comment (hachreak) [inline invenio_deposit/api.py]: > lca = published.revisions[self['_deposit']['pid']['revision']]

as `published` do you mean `self`?

Comment (jirikuncar) [inline invenio_deposit/api.py]: no I mean `published`

Comment (hachreak) [inline invenio_deposit/api.py]: ah, ok.. now I understand. The `published` was from you first comment..
I didn't understand because in the code you wrote the `published` variable was renamed in `first` when defined but one line after you refer to `published`.

File: invenio_deposit/api.py [hunk 8]
```diff
@@ -211,12 +218,16 @@ def process_files(data):
...
             record_pid, record = self.fetch_published()
-            # TODO add support for patching
-            assert record.revision_id == self['_deposit']['pid']['revision_id']
+            if record.revision_id == self['_deposit']['pid']['revision_id']:
+                data = dict(self.dumps())
+            else:
+                data = self.merge_with_published()
 
-            data = dict(self.dumps())
+            data['$schema'] = schema
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: Why?

File: invenio_deposit/api.py [hunk 9]
```diff
@@ -131,12 +134,16 @@ def fetch_published(self):
...
-        # TODO if revisions are the same then we can apply directly
-        # latest = self.fetch_published()
-        # initial = A.revisions[self['_deposit']['pid']['revision']]
-        # patch = diff(initial, self)
-        # latest.apply(patch)
+        """Merge changes with latest published version."""
+        pid, first = self.fetch_published()
+        lca = first.revisions[self['_deposit']['pid']['revision_id']]
+        second = self
+        m = Merger(lca.dumps(), first.dumps(), second.dumps(), {})
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: We should probably remove/ignore `_deposit` and `$schema` fields as they will be always different. The`_deposit` field should probably be set after merge like https://github.com/inveniosoftware/invenio-deposit/pull/63/files#diff-8d60a8f35673846c2c55cf5cde4c9b85R231.

File: invenio_deposit/api.py (hunk 8 above)
Comment (hachreak) [inline invenio_deposit/api.py]: > Please note that we need to take care about $schema and _deposit fields.

Did you say to take care about $schema and I though I should preserve like you did for `_deposit` field here: https://github.com/inveniosoftware/invenio-deposit/blob/master/invenio_deposit/api.py#L306-L308
Should I do in a different way?

Comment (jirikuncar) [inline invenio_deposit/api.py]: Yes, that's done by `data['$schema'] = self.record_schema` on next line.

Comment (jirikuncar) [inline invenio_deposit/api.py]: The `_deposit` and `$schema` should be "preserved" in `merge_with_published`.

File: invenio_deposit/errors.py [hunk 10]
```diff
@@ -37,7 +37,14 @@ class FileAlreadyExists(RESTException):
...
 
     code = 400
     description = 'Wrong file on input.'
+
+
+class MergeConflict(RESTException):
+    """Error on merging a deposit."""
+
+    code = 409
+    description = 'Deposit merge conflicts'
```
Comment (jirikuncar) [inline invenio_deposit/errors.py]: missing dot `.`

File: tests/conftest.py [hunk 11]
```diff
@@ -81,14 +81,15 @@ def app(request):
...
-            'SQLALCHEMY_DATABASE_URI', 'sqlite://'),
+            'SQLALCHEMY_DATABASE_URI', 'sqlite:///test.db'),
         SQLALCHEMY_TRACK_MODIFICATIONS=True,
         TESTING=True,
         WTF_CSRF_ENABLED=False,
         SECURITY_PASSWORD_HASH='plaintext',
         SECURITY_PASSWORD_SCHEMES=['plaintext'],
         OAUTHLIB_INSECURE_TRANSPORT=True,
         OAUTH2_CACHE_TYPE='simple',
+        #  SQLALCHEMY_ECHO=True,
```
Comment (jirikuncar) [inline tests/conftest.py]: 💩

File: invenio_deposit/api.py (hunk 9 above)
Comment (hachreak) [inline invenio_deposit/api.py]: do you mean like this?
```diff
diff --git a/invenio_deposit/api.py b/invenio_deposit/api.py
index 82547cf..6033636 100644
--- a/invenio_deposit/api.py
+++ b/invenio_deposit/api.py
@@ -138,10 +138,20 @@ class Deposit(Record):
         pid, first = self.fetch_published()
         lca = first.revisions[self['_deposit']['pid']['revision_id']]
         second = self
-        m = Merger(lca.dumps(), first.dumps(), second.dumps(), {})
+        # ignore _deposit and $schema field
+        lca_dict = lca.dumps()
+        first_dict = first.dumps()
+        second_dict = second.dumps()
+        lca_dict.pop('$schema')
+        lca_dict.pop('_deposit')
+        first_dict.pop('$schema')
+        first_dict.pop('_deposit')
+        second_dict.pop('$schema')
+        second_dict.pop('_deposit')
+        m = Merger(lca_dict, first_dict, second_dict, {})
         try:
             m.run()
```

Comment (jirikuncar) [inline invenio_deposit/api.py]:
```python
args = [rev.dumps() for rev in (lca, first, second)]
for rev in args:
    del rev['_deposit'], rev['$schema']
args.append({})
m = Merger(*args)
```

**Comment by hachreak** [comment]
ping @lnielsen
